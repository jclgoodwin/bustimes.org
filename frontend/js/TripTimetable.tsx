import React, { type ReactElement } from "react";
import type { StopTime, VehicleJourney } from "./JourneyMap";
import type { Vehicle } from "./VehicleMarker";

export type TripTime = {
  id: number;
  stop: {
    name: string;
    atco_code?: string;
    location?: [number, number];
    icon?: string | null;
    bearing?: number | null;
  };
  track?: [number, number][] | null;
  aimed_arrival_time: string | null;
  aimed_departure_time: string | null;
  expected_arrival_time?: string | null;
  expected_departure_time?: string | null;
  actual_departure_time?: string;
  // actual_arrival_time: string;
  timing_status: string;
  pick_up?: boolean;
  set_down?: boolean;
  call_condition?: string | null;
};

type Note = {
  code: string;
  text: string;
};

export type Trip = {
  id?: number;
  vehicle_journey_code?: string;
  ticket_machine_code?: string;
  block?: string;
  service?: {
    slug?: string;
    id: number | null;
    line_name?: string;
    mode?: string;
  };
  operator?: {
    slug?: string;
    noc: string;
    name: string;
    vehicle_mode: string;
  };
  times: TripTime[];
  notes?: Note[];
};

function Row({
  stop,
  onMouseEnter,
  vehicle,
  aimedColumn,
  highlightedStop,
  first = false,
  last = false,
}: {
  stop: TripTime;
  onMouseEnter?: (stop: TripTime) => void;
  vehicle?: Vehicle;
  aimedColumn?: boolean;
  highlightedStop?: string;
  first: boolean;
  last: boolean;
}) {
  const handleMouseEnter = React.useCallback(() => {
    if (onMouseEnter) {
      if (stop.stop.location) {
        onMouseEnter(stop);
      }
    }
  }, [stop, onMouseEnter]);

  let className: string | undefined;

  let stopName: string | ReactElement = stop.stop.name;
  if (stop.stop.icon) {
    stopName = `${stopName} (${stop.stop.icon})`;
  }
  if (stop.stop.atco_code) {
    const url = `/stops/${stop.stop.atco_code}`;
    if (url === highlightedStop) {
      className = "is-highlighted";
    }
    stopName = <a href={url}>{stopName}</a>;
  }

  if (stop.timing_status && stop.timing_status !== "PTP") {
    className = className ? `${className} minor` : "minor";
  }

  let rowSpan: number | undefined;
  if (
    aimedColumn &&
    stop.aimed_arrival_time &&
    stop.aimed_departure_time &&
    stop.aimed_arrival_time !== stop.aimed_departure_time
  ) {
    rowSpan = 2;
  }

  let actual: string | null | ReactElement | undefined;
  let actualRowSpan = rowSpan;

  actual = stop.expected_departure_time || stop.expected_arrival_time; // Irish live departures

  if (!actual) {
    if (vehicle?.progress && vehicle.progress.id === stop.id) {
      actual = vehicle.datetime;
      if (vehicle.progress.progress > 0.1) {
        actualRowSpan = (actualRowSpan || 1) + 1;
      }
    } else if (!vehicle?.progress || vehicle.progress.id > stop.id) {
      actual = stop.actual_departure_time; // vehicle history
    }
    if (actual) {
      actual = actual.slice(11, 16);
    }
  }
  if (actual) {
    actual = <td rowSpan={actualRowSpan}>{actual}</td>;
  }

  let caveat: ReactElement | undefined;
  if (!first && !last) {
    if (stop.set_down === false) {
      if (stop.pick_up === false) {
        caveat = <abbr title="does not stop">pass</abbr>;
      } else {
        caveat = <abbr title="picks up only">p</abbr>;
      }
    } else if (stop.pick_up === false) {
      caveat = <abbr title="sets down only">s</abbr>;
    }
  }

  let aimed: ReactElement | null | string = null;
  if (aimedColumn) {
    aimed = stop.aimed_arrival_time || stop.aimed_departure_time;
    if (stop.call_condition === "notStopping") {
      aimed = (
        <td>
          <s>
            {aimed}
            {caveat}
          </s>
        </td>
      );
    } else {
      aimed = (
        <td>
          {aimed}
          {caveat}
        </td>
      );
    }
  }

  return (
    <React.Fragment>
      <tr className={className} onMouseEnter={handleMouseEnter}>
        <td className="stop-name" rowSpan={rowSpan}>
          {stopName}
        </td>
        {aimed}
        {actual}
      </tr>
      {rowSpan ? (
        <tr className={className}>
          <td>{stop.aimed_departure_time}</td>
        </tr>
      ) : null}
    </React.Fragment>
  );
}

export const tripFromJourney = (journey: VehicleJourney): Trip | undefined => {
  if (journey.stops) {
    return {
      times: journey.stops.map((stop, i: number) => {
        return {
          id: stop.id,
          stop: {
            atco_code: stop.atco_code,
            name: stop.name,
            location: stop.coordinates || undefined,
          },
          timing_status: stop.minor ? "OTH" : "PTP",
          aimed_arrival_time: stop.aimed_arrival_time,
          aimed_departure_time: stop.aimed_departure_time,
          actual_departure_time: stop.actual_departure_time,
        };
      }),
    };
  }
};

const TripTimetable = React.memo(function TripTimetable({
  trip,
  onMouseEnter,
  vehicle,
  highlightedStop,
}: {
  trip: Trip;
  onMouseEnter?: (stop: TripTime) => void;
  vehicle?: Vehicle;
  highlightedStop?: string;
}) {
  const [showEarlierStops, setShowEarlierStops] = React.useState(false);

  const aimedColumn = trip.times?.some(
    (item: TripTime) => item.aimed_arrival_time || item.aimed_departure_time,
  );
  const actualColumn =
    vehicle ||
    trip.times?.some(
      (item: TripTime) =>
        item.actual_departure_time ||
        item.expected_arrival_time ||
        item.expected_departure_time,
    );

  let earlierStops = false;

  let times = trip.times;
  if (!showEarlierStops && vehicle && vehicle.progress) {
    const index = times.findIndex((item) => item.id === vehicle.progress?.id);
    if (index > 0) {
      times = times.slice(index);
      earlierStops = true;
    }
  }
  const indexOfLastRow = times.length - 1;

  return (
    <React.Fragment>
      {earlierStops || showEarlierStops ? (
        <label>
          <input
            type="checkbox"
            checked={showEarlierStops}
            onChange={() => setShowEarlierStops(!showEarlierStops)}
          />
          {" Show previous stops"}
        </label>
      ) : null}
      <table>
        <thead>
          <tr>
            <th className="stop-name" />
            {aimedColumn ? <th>Sched&shy;uled</th> : null}
            {actualColumn ? <th>Actual</th> : null}
          </tr>
        </thead>
        <tbody>
          {times.map((stop, i) => (
            <Row
              key={stop.id || i}
              aimedColumn={aimedColumn}
              stop={stop}
              onMouseEnter={onMouseEnter}
              vehicle={vehicle}
              highlightedStop={highlightedStop}
              first={i === 0 && !earlierStops}
              last={i === indexOfLastRow}
            />
          ))}
        </tbody>
      </table>
      {trip.notes?.map((note) => (
        <p key={note.code}>{note.text}</p>
      ))}
    </React.Fragment>
  );
});

export default TripTimetable;
