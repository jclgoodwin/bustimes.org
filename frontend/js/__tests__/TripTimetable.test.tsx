/**
 * @jest-environment jsdom
 */

import { render, screen } from "@testing-library/react";
import TripTimetable, { type Trip } from "../TripTimetable";
import type { Vehicle } from "../VehicleMarker";

const trip: Trip = {
  id: 273819070,
  vehicle_journey_code: "vj_41",
  ticket_machine_code: "29",
  block: "",
  service: { id: 4441, line_name: "502", mode: "bus" },
  operator: { noc: "KCTB", name: "Konectbus", vehicle_mode: "bus" },
  notes: [],
  times: [
    {
      id: 16494874561,
      stop: {
        atco_code: "2900C0305",
        name: "Harford Park & Ride",
        location: [1.2720919815463243, 52.588796929212215],
        bearing: 135,
        icon: null,
      },
      aimed_arrival_time: null,
      aimed_departure_time: "10:05",
      track: null,
      timing_status: "PTP",
      pick_up: true,
      set_down: false,
      expected_arrival_time: null,
      expected_departure_time: null,
    },
    {
      id: 16494874562,
      stop: {
        atco_code: "2900N1237",
        name: "County Hall Bus Shelter",
        location: [1.3073762895735876, 52.61532759989593],
        bearing: 225,
        icon: "A",
      },
      aimed_arrival_time: null,
      aimed_departure_time: "10:13",
      track: null,
      timing_status: "PTP",
      pick_up: true,
      set_down: true,
      expected_arrival_time: null,
      expected_departure_time: null,
    },
    {
      id: 16494874563,
      stop: {
        atco_code: "2900N12909",
        name: "Norwich Bus Station",
        location: [1.292518141141929, 52.6240700123112],
        bearing: null,
        icon: "F",
      },
      aimed_arrival_time: "10:18",
      aimed_departure_time: "10:20",
      track: null,
      timing_status: "PTP",
      pick_up: true,
      set_down: true,
      expected_arrival_time: null,
      expected_departure_time: null,
    },
    {
      id: 16494874564,
      stop: {
        atco_code: "2900N12245",
        name: "Norwich Tombland",
        location: [1.298782630795197, 52.63091962080655],
        bearing: 0,
        icon: "CM",
      },
      aimed_arrival_time: null,
      aimed_departure_time: "10:24",
      track: null,
      timing_status: "OTH",
      pick_up: true,
      set_down: true,
      expected_arrival_time: null,
      expected_departure_time: null,
    },
    {
      id: 16494874565,
      stop: {
        atco_code: "2900N12233",
        name: "Norwich Anglia Square",
        location: [1.2963982671391574, 52.63577066440402],
        bearing: 0,
        icon: "B",
      },
      aimed_arrival_time: null,
      aimed_departure_time: "10:27",
      track: null,
      timing_status: "PTP",
      pick_up: true,
      set_down: true,
      expected_arrival_time: null,
      expected_departure_time: null,
    },
    {
      id: 16494874566,
      stop: {
        atco_code: "2900S32517",
        name: "Sprowston Park and Ride",
        location: [1.33416997295962, 52.66492263745109],
        bearing: 225,
        icon: null,
      },
      aimed_arrival_time: "10:40",
      aimed_departure_time: null,
      track: null,
      timing_status: "PTP",
      pick_up: true,
      set_down: true,
      expected_arrival_time: null,
      expected_departure_time: null,
    },
  ],
};

const vehicle: Vehicle = {
  id: 20158,
  coordinates: [1.308246, 52.617555],
  heading: 303,
  datetime: "2023-09-15T09:17:19Z",
  destination: "Sprowston",
  block: "601067",
  trip_id: 273819070,
  service_id: 4441,
  service: {
    url: "/services/502-park-ride-harford-pr-norwich-city-centre-sprow",
    line_name: "502",
  },
  vehicle: {
    url: "/vehicles/kctb-goea-630",
    name: "630 - SN65 OAL",
    features: "Double decker",
    livery: 405,
    colour: "#79b33b",
  },
  progress: {
    id: 16494874562,
    sequence: 2,
    prev_stop: "2900N1237",
    next_stop: "2900N12909",
    progress: 0.022,
  },
  delay: 252,
};

it("renders trip timetable", async () => {
  const { container, rerender } = render(<TripTimetable trip={trip} />);

  expect(container).toMatchSnapshot();

  rerender(<TripTimetable trip={trip} vehicle={vehicle} />);
  expect(container).toMatchSnapshot();

  rerender(
    <TripTimetable
      trip={trip}
      vehicle={vehicle}
      highlightedStop="/stops/2900N12245"
    />,
  );
  expect(container).toMatchSnapshot();
});
