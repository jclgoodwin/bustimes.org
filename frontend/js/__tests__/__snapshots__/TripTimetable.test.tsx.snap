// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders trip timetable 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          class="stop-name"
        />
        <th>
          Sched­uled
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900C0305"
          >
            Harford Park & Ride
          </a>
        </td>
        <td>
          10:05
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N1237"
          >
            County Hall Bus Shelter (A)
          </a>
        </td>
        <td>
          10:13
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
          rowspan="2"
        >
          <a
            href="/stops/2900N12909"
          >
            Norwich Bus Station (F)
          </a>
        </td>
        <td>
          10:18
        </td>
      </tr>
      <tr>
        <td>
          10:20
        </td>
      </tr>
      <tr
        class="minor"
      >
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N12245"
          >
            Norwich Tombland (CM)
          </a>
        </td>
        <td>
          10:24
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N12233"
          >
            Norwich Anglia Square (B)
          </a>
        </td>
        <td>
          10:27
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900S32517"
          >
            Sprowston Park and Ride
          </a>
        </td>
        <td>
          10:40
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders trip timetable 2`] = `
<div>
  <label>
    <input
      type="checkbox"
    />
     Show previous stops
  </label>
  <table>
    <thead>
      <tr>
        <th
          class="stop-name"
        />
        <th>
          Sched­uled
        </th>
        <th>
          Actual
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N1237"
          >
            County Hall Bus Shelter (A)
          </a>
        </td>
        <td>
          10:13
        </td>
        <td>
          09:17
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
          rowspan="2"
        >
          <a
            href="/stops/2900N12909"
          >
            Norwich Bus Station (F)
          </a>
        </td>
        <td>
          10:18
        </td>
      </tr>
      <tr>
        <td>
          10:20
        </td>
      </tr>
      <tr
        class="minor"
      >
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N12245"
          >
            Norwich Tombland (CM)
          </a>
        </td>
        <td>
          10:24
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N12233"
          >
            Norwich Anglia Square (B)
          </a>
        </td>
        <td>
          10:27
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900S32517"
          >
            Sprowston Park and Ride
          </a>
        </td>
        <td>
          10:40
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders trip timetable 3`] = `
<div>
  <label>
    <input
      type="checkbox"
    />
     Show previous stops
  </label>
  <table>
    <thead>
      <tr>
        <th
          class="stop-name"
        />
        <th>
          Sched­uled
        </th>
        <th>
          Actual
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N1237"
          >
            County Hall Bus Shelter (A)
          </a>
        </td>
        <td>
          10:13
        </td>
        <td>
          09:17
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
          rowspan="2"
        >
          <a
            href="/stops/2900N12909"
          >
            Norwich Bus Station (F)
          </a>
        </td>
        <td>
          10:18
        </td>
      </tr>
      <tr>
        <td>
          10:20
        </td>
      </tr>
      <tr
        class="is-highlighted minor"
      >
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N12245"
          >
            Norwich Tombland (CM)
          </a>
        </td>
        <td>
          10:24
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900N12233"
          >
            Norwich Anglia Square (B)
          </a>
        </td>
        <td>
          10:27
        </td>
      </tr>
      <tr>
        <td
          class="stop-name"
        >
          <a
            href="/stops/2900S32517"
          >
            Sprowston Park and Ride
          </a>
        </td>
        <td>
          10:40
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
